# AUTOMATIC TICKET ASSIGNMENT

### Milestone 1 : Data pre-processing and EDA

Importing Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder,scale

In [2]:
import os

In [4]:
data='DATA'
raw_data_folder='RAW_DATA'
processed_data_folder='PROCESSED_DATA'

In [5]:
if data not in os.listdir(os.getcwd()):
    os.makedirs(os.path.join(data))
DATA_PATH=os.path.join(data)
if raw_data_folder not in os.listdir(os.path.join(DATA_PATH)):
    os.makedirs(os.path.join(DATA_PATH,raw_data_folder))
RAW_DATA_PATH=os.path.join(DATA_PATH,raw_data_folder)
if processed_data_folder not in os.listdir(os.path.join(DATA_PATH)):
    os.makedirs(os.path.join(DATA_PATH,processed_data_folder))
PROCESSED_DATA_PATH=os.path.join(DATA_PATH,processed_data_folder)

Read the csv file and see some records

In [6]:
new_df=pd.read_csv(os.path.join(RAW_DATA_PATH,'Input Data Translated.csv'), encoding = 'latin-1')
new_df.head()

Short description  \
0                    login issue   
1                        outlook   
2             cant log in to vpn   
3  unable to access hr_tool page   
4                    skype error   

                                         Description             Caller  \
0  -verified user details.(employee# & manager na...  spxjnwir pjlcoqds   
1  \n\nreceived from: hmjdrvpb.komuaywn@gmail.com...  hmjdrvpb komuaywn   
2  \n\nreceived from: eylqgodm.ybqkwiam@gmail.com...  eylqgodm ybqkwiam   
3                      unable to access hr_tool page  xbkucsvz gcpydteq   
4                                        skype error  owlgqjme qhcozdfx   

  Assignment group Lang_Short_Description Lang_Description  
0            GRP_0                     en               en  
1            GRP_0                     en               en  
2            GRP_0                     en               en  
3            GRP_0                     en               en  
4            GRP_0                     en               en

In [7]:
new_df.shape

(8500, 6)

There are 8500 rows and 4 columns

## EDA

Understanding of the columns is important in order to understand whether to retain/ drop some of them.

# Caller Column

This perhaps indicates the person who filed the ticket; it would be useful to understand the number of such callers (If there are many unique callers with low sample bases, then it might not make sense to include this column

In [104]:
new_df.Caller.nunique() # Understanding the unique no. of callers

2950

Thus across 8500 Rows, there are 2950 unique callers and hence each caller is unlikely to have a sufficient base for analysis. However would be interesting to check the counts of at least the callers who put in tickets more frequently

In [105]:
new_df.Caller.value_counts()

bpctwhsn kzqsbmtp    810
ZkBogxib QsEJzdZO    151
fumkcsji sarmtlhy    134
rbozivdq gmlhrtvp     87
rkupnshb gsmzfojw     71
jloygrwh acvztedi     64
spxqmiry zpwgoqju     63
oldrctiu bxurpsyi     57
olckhmvx pcqobjnd     54
dkmcfreg anwmfvlg     51
jyoqwxhz clhxsoqy     51
efbwiadp dicafxhv     45
afkstcev utbnkyop     32
gzhapcld fdigznbk     30
mnlazfsr mtqrkhnx     28
uvrbhlnt bjrmalzi     27
entuakhp xrnhtdmk     25
jionmpsf wnkpzcmv     24
vzqomdgt jwoqbuml     24
bozdftwx smylqejw     23
rxoynvgi ntgdsehl     21
qasdhyzm yuglsrwx     21
utyeofsk rdyzpwhi     21
vbwszcqn nlbqsuyv     19
ugyothfz ugrmkdhx     17
hbmwlprq ilfvyodx     17
vfrdxtqw jfbmsenz     16
niptbwdq csenjruz     16
pfzxecbo ptygkvzl     16
ughzilfm cfibdamq     16
                    ... 
zkhmvnib sugzyvhp      1
xaykwtrf amlswjfr      1
atdclmyi wqxzaysu      1
szcbhvwe edpouqjl      1
sagfvdeb kfqtvgxl      1
puxyqwjn zgmnsvjx      1
cwryvksu cedsairg      1
cvnzhbfa qcvzyrgo      1
suyighpt wifujdvq      1


There are only 11 callers who have a sample base > 50; While the top caller has 810 records, the sample per caller diminishes rapidly; thus there does not seem to be any value in retaining this column and it will be dropped 

Let's see the description

In [107]:
new_df.describe().transpose()

count unique                top  freq
Short description  8498   7386     password reset    48
Description        8497   7751                the    56
Caller             8500   2950  bpctwhsn kzqsbmtp   810
Assignment group   8500     74              GRP_0  3976
Language           8497     27                 en  7733

Above description says below observations:
1. 'Short description' column has 2 blank values and 'Desciption' column has 3 blank values
2. Out of 8498 there are 7386 unique values in 'Short description' column, out of 8497 there are 7751 unique values in 'Description' column and out of 8500 there are 74 unique values in 'Assignment group' column which is target column. Also there are 27 unique languages in the database
3. "Password reset" is the top frequent value with 48 occurrences in 'Short description' column, "the" is the top frequent value with 56 occurrences in 'Description' column and "GRP_0" is the group which has maximum assignment of tickets i.e 3976. English is the most common language with 7733 records

Comments: In the Description Column 'the' by itself does not add any value since it would get removed as a stop word; hence useful to see what is Short Description that is in the rows with "the" in the Description Column

In [108]:
new_df[new_df.Description == 'the'].head(10)

Short description Description  \
1049  reset passwords for soldfnbq uhnbsvqd using pa...         the   
1054  reset passwords for fygrwuna gomcekzi using pa...         the   
1144  reset passwords for wvdxnkhf jirecvta using pa...         the   
1184  reset passwords for pxvjczdt kizsjfpq using pa...         the   
1292  reset passwords for cubdsrml znewqgop using pa...         the   
1476  reset passwords for bnoupaki cpeioxdz using pa...         the   
1558  reset passwords for usa feathers using passwor...         the   
1693  reset passwords for eglavnhx uprodleq using pa...         the   
1834  reset passwords for hybiaxlk lawptzir using pa...         the   
1850  reset passwords for fylrosuk kedgmiul using pa...         the   

                 Caller Assignment group Language  
1049  soldfnbq uhnbsvqd           GRP_17       en  
1054  fygrwuna gomcekzi           GRP_17       en  
1144  wvdxnkhf jirecvta           GRP_17       en  
1184  pxvjczdt kizsjfpq           GRP_17       en  
1292  cubdsrml znewqgop           GRP_17       en  
1476  bnoupaki cpeioxdz           GRP_17       en  
1558  lmqysdec ljvbnpqw           GRP_17       en  
1693  eglavnhx uprodleq           GRP_17       en  
1834  hybiaxlk lawptzir           GRP_17       en  
1850  fylrosuk kedgmiul           GRP_17       en

Most of the "the" in Description seem to be associated with "reset passwords" in the Short Description Column; thus in order to develop a reasonable model it points to the fact that it would be good to concatenate both these columns

Checking for the Data Types

In [109]:
new_df.dtypes

Short description    object
Description          object
Caller               object
Assignment group     object
Language             object
dtype: object

Let's see number of tickets assigned to each group

In [112]:
print(new_df['Assignment group'].value_counts())

GRP_0     3976
GRP_8      661
GRP_24     289
GRP_12     257
GRP_9      252
GRP_2      241
GRP_19     215
GRP_3      200
GRP_6      184
GRP_13     145
GRP_10     140
GRP_5      129
GRP_14     118
GRP_25     116
GRP_33     107
GRP_4      100
GRP_29      97
GRP_18      88
GRP_16      85
GRP_17      81
GRP_31      69
GRP_7       68
GRP_34      62
GRP_26      56
GRP_40      45
GRP_28      44
GRP_41      40
GRP_15      39
GRP_30      39
GRP_42      37
          ... 
GRP_44      15
GRP_36      15
GRP_50      14
GRP_53      11
GRP_65      11
GRP_52       9
GRP_55       8
GRP_51       8
GRP_59       6
GRP_49       6
GRP_46       6
GRP_43       5
GRP_32       4
GRP_66       4
GRP_38       3
GRP_56       3
GRP_58       3
GRP_68       3
GRP_63       3
GRP_54       2
GRP_72       2
GRP_71       2
GRP_57       2
GRP_69       2
GRP_73       1
GRP_61       1
GRP_67       1
GRP_64       1
GRP_70       1
GRP_35       1
Name: Assignment group, Length: 74, dtype: int64


The Group variable has a very long tail and it might not be beneficial to keep the ones that have a low record base; One could potentially club the groups with low record base into 1 group called as "Others". Classic statistics state that beyond a sample of 30, distributions tend to be normal and thus this value could be set as a cut-off for a group to be included and all groups with a base of <30 can be combined into "Others"

Let's check for null values

In [113]:
new_df.isnull().apply(pd.value_counts)  

Short description  Description  Caller  Assignment group  Language
False               8498         8497  8500.0            8500.0      8497
True                   2            3     NaN               NaN         3

There are 2 null values in 'Short description' column and 3 in 'Description' column.There is no null value in target column.

If there are rows, with Null values for both, then we would need to remove them; however we need to check for the same

In [114]:
# Let's look at the rows with null values
new_df[pd.isnull(new_df).any(axis=1)]

Short description  \
2604                               NaN   
3383                               NaN   
4395          i am locked out of skype   
6371  authorization add/delete members   
7397                   browser issue :   

                                            Description             Caller  \
2604  \n\nreceived from: ohdrnswl.rezuibdt@gmail.com...  ohdrnswl rezuibdt   
3383  \n-connected to the user system using teamview...  qftpazns fxpnytmk   
4395                                                NaN  viyglzfo ajtfzpkb   
6371                                                NaN  hpmwliog kqtnfvrl   
7397                                                NaN  fgejnhux fnkymoht   

     Assignment group Language  
2604           GRP_34       en  
3383            GRP_0       en  
4395            GRP_0      NaN  
6371            GRP_0      NaN  
7397            GRP_0      NaN

There are no records which have Null values for both Short Description and Description; Given that for prediction it is likely that both Columns would be used, replacement of Null values can be done with a blank string (as it is likely that both Short Description and Description would be concatenated while building the model)

In [115]:
# NULL replacement
new_df.fillna(str(), inplace=True)
new_df[pd.isnull(new_df).any(axis=1)]

Empty DataFrame
Columns: [Short description, Description, Caller, Assignment group, Language]
Index: []

Now there is no null value in any column

Checking for Duplicates across Short Description and Description

In [116]:
df_common=new_df[new_df[["Short description","Description"]].apply(lambda x : x[0]==x[1],axis=1)].reset_index(drop=True).copy()

In [117]:
df_common.head()

Short description  \
0                      unable to access hr_tool page   
1                                        skype error   
2     unable to log in to engineering tool and skype   
3  ticket_no1550391- employment status - new non-...   
4               unable to disable add ins on outlook   

                                         Description             Caller  \
0                      unable to access hr_tool page  xbkucsvz gcpydteq   
1                                        skype error  owlgqjme qhcozdfx   
2     unable to log in to engineering tool and skype  eflahbxn ltdgrvkz   
3  ticket_no1550391- employment status - new non-...  eqzibjhw ymebpoih   
4               unable to disable add ins on outlook  mdbegvct dbvichlg   

  Assignment group Language  
0            GRP_0       en  
1            GRP_0       en  
2            GRP_0       en  
3            GRP_0       en  
4            GRP_0       en

In [118]:
df_common.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2889 entries, 0 to 2888
Data columns (total 5 columns):
Short description    2889 non-null object
Description          2889 non-null object
Caller               2889 non-null object
Assignment group     2889 non-null object
Language             2889 non-null object
dtypes: object(5)
memory usage: 112.9+ KB


Comments: this shows that out of the 8500 records, 2889 have the same information in both Short description and Description - This would point out that it is better to concatenate the two columns for the modeling stage

Language as a column is not required for the modeling but was just for understanding; dropping the Language Column

In [119]:
new_df=new_df.drop('Language',axis=1)
new_df.shape
new_df.head()

Short description  \
0                    login issue   
1                        outlook   
2             cant log in to vpn   
3  unable to access hr_tool page   
4                    skype error   

                                         Description             Caller  \
0  -verified user details.(employee# & manager na...  spxjnwir pjlcoqds   
1  \n\nreceived from: hmjdrvpb.komuaywn@gmail.com...  hmjdrvpb komuaywn   
2  \n\nreceived from: eylqgodm.ybqkwiam@gmail.com...  eylqgodm ybqkwiam   
3                      unable to access hr_tool page  xbkucsvz gcpydteq   
4                                        skype error  owlgqjme qhcozdfx   

  Assignment group  
0            GRP_0  
1            GRP_0  
2            GRP_0  
3            GRP_0  
4            GRP_0

# Data Understanding thus far

1. All text has been converted into English
2. Need to concatenate both the short Description and Description Columns
3. Long tails in "Assignment group" column need to be taken care of
4. Null values solved for by replacing with blank string
5. Need to remove email id's from text